[mandykqh] iP

mandykqh opened the pull request:
# Genie
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

GeniePro frees your mind of having to remember things you need to do. It's,

* GUI-based,
* easy to learn
* ~~FAST~~ *SUPER FAST* to use
* All you need to do is,

download the most recent release of  genie.jar from [here](https://github.com/mandykqh/ip).
1. run genie.jar
2. add your tasks.
3. let it manage your tasks for you 😉
And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Reminders (coming soon)

If you are a Java programmer, here's the ```main``` method that launches Genie:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

mandykqh committed:
Complete Level 1 - Greet, Echo, Exit

mandykqh committed:
Complete Level 2 - Add, List

mandykqh committed:
Refactor code for Level 2

mandykqh committed:
Complete Level 3 - Mark as Done

mandykqh committed:
Complete Level 4 - ToDos, Events, Deadlines

mandykqh committed:
Complete A-TextUiTesting - Automated Text UI Testing

mandykqh committed:
Complete Level 5 - Handle Errors

mandykqh committed:
Complete Level 6 - Delete

mandykqh committed:
Add .gitignore for ./data/duke.txt

mandykqh committed:
Implement Level 7 - Save

mandykqh committed:
Implement Level 8 - Dates and Times

mandykqh committed:
Merge branch 'branch-Level-7'

mandykqh committed:
Merge branch 'branch-Level-8'

mandykqh committed:
Refactor code for recognising user input command

mandykqh committed:
Revert "Refactor code for recognising user input command"

This reverts commit b4c03d78ce884231e06bddce7b7f8bfc72ca1c34.

mandykqh committed:
Refactor code and handle more cases of EmptyInputException

mandykqh committed:
Make the code more OOP

mandykqh committed:
Fix bugs and personalise bot UI

mandykqh committed:
Implement A-MoreOOP

mandykqh committed:
Implement A-Packages - Organize into Packages

mandykqh committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

mandykqh committed:
Implement A-JUnit: Add JUnit Tests

mandykqh committed:
Implement A-JavaDoc: JavaDoc

mandykqh committed:
Implement A-CodingStandard: Follow the Coding Standard

mandykqh committed:
Implement Level-9: Find

mandykqh committed:
Merge branch 'branch-A-JavaDoc'

mandykqh committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/genie/Storage.java
#	src/main/java/genie/TaskList.java
#	src/main/java/genie/Ui.java

mandykqh committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/genie/task/Task.java

mandykqh committed:
Fix A-TextUiTesting

heeeyi commented on src/main/java/genie/Parser.java:
```diff
@@ -0,0 +1,66 @@
+     * @return <Code>Command</Code> aligning to the user's input
+     * @throws DukeException if error occurs in parsing
+     */
+    public Command parse(String in) throws DukeException {
```
This method is a bit too long (above 30 LOC), is it possible to shorten it?

heeeyi commented on src/main/java/genie/Parser.java:
```diff
@@ -0,0 +1,66 @@
+            }
+        }
+    }
+    public boolean isACommand(String i) {
```
This expression is a bit complicated, could consider using if-else?

heeeyi commented on src/main/java/genie/Ui.java:
```diff
@@ -0,0 +1,166 @@
+     * Prints add farewell message.
+     */
+    public void showFarewellMessage() {
+        System.out.println("Bye. Hope to see you again soon!");
```
This is a bit of a magic string, consider adding it to the static attributes?

heeeyi commented on src/main/java/genie/Ui.java:
```diff
@@ -0,0 +1,166 @@
+     * @param size number of tasks in the list
+     */
+    public void showAddTaskMessage(Task t, int size) {
+        System.out.println("Got it. I've added this task:\n  " + t.toString());
```
Try to follow the "SLAP Hard" principle, and make the other 2 printing as methods as well?

heeeyi reviewed (commented):
In overall, nice code that is clean, easy to read and follows the codings standard! Keep it up!

ShanHng commented on src/main/java/genie/Genie.java:
```diff
@@ -4,6 +4,7 @@
 import genie.exception.DukeException;
 
 import java.io.*;
```
Perhaps this should be split into single-line import statements?

ShanHng commented on src/main/java/genie/Storage.java:
```diff
@@ -15,14 +15,14 @@
 public class Storage {
     private static final String PATH = "./data/duke.txt";
     private static FileWriter fw;
```
Perhaps we could use a longer variable name here, since the scope in which this FileWriter is used is quite large?

ShanHng commented on src/main/java/genie/task/Deadline.java:
```diff
@@ -32,7 +32,7 @@ public String toString() {
             String formattedDate = new SimpleDateFormat("EEE, dd/MM/yyyy, h:mm a").format(date);
             this.by = formattedDate;
```
If this is a field, maybe we could use a more descriptive name here for "by"?

ShanHng commented on src/main/java/genie/task/Task.java:
```diff
@@ -1,30 +1,63 @@
+    /**
+     * Changes the status of the task to marked undone.
+     */
     public void unmarkDone() {
```
Just a small nit, but would "markUndone" sound more natural and easier to remember?

ShanHng commented on src/main/java/genie/task/Event.java:
```diff
@@ -2,20 +2,39 @@
+ * Deals with creating event tasks.
+ */
 public class Event extends Task {
     protected String from;
```
Perhaps more descriptive names can be used for these fields that store dates instead of just "to"and "from"?

ShanHng commented on src/main/java/genie/TaskList.java:
```diff
@@ -47,13 +82,23 @@ public Task addDeadlineFromFile(String sf) {
+     */
     public Task addEventFromUser(String i) {
         String[] contents = i.split(" /from ");
         String[] fromTo = contents[1].split(" /to ");
```
Since this is an array, perhaps we could use a name that can indicate that?

ShanHng commented on src/main/java/genie/task/Task.java:
```diff
@@ -28,5 +31,16 @@ public String getStatusBox() {
+        return description;
+    }
+    public boolean containsWord(String keyword) {
+        Pattern pattern = Pattern.compile("\\b" + keyword + "\\b", Pattern.CASE_INSENSITIVE);
```
Love that you are using a regex pattern to make Task "recognize" words on their own!

ShanHng reviewed (commented):
LGTM. Generally quite easy to look through! Just some minor things that may benefit from some small change! 😃

mandykqh committed:
Implement Level-10: GUI

mandykqh committed:
Merge branch 'branch-Level-10'

mandykqh committed:
Enable assertions in Gradle

mandykqh committed:
Add Genie greeting message

The application boots up a blank window that waits for user's input.

Receiving Genie's greeting first gives a more 'complete' appearance and
ensures that the application has started successfully.

Follows weak SLAP principle as there is no abstraction of the greeting
message due to FXML load exception when done so.

mandykqh committed:
Implement A-Assertions

There are assumptions in the values passed in methods that are not
defined.

Assertions are used to define these assumptions.

Assertion failure can help detect bugs in the code at runtime.

mandykqh committed:
Move classes into main package

mandykqh committed:
Implement A-CodeQuality

Some code does not follow good code quality.

Refactor the code to improve code quality, such as minimizing
complicated expressions, magic numbers, renaming variables, following
SLAP principle, etc.

This helps to maximize readability for other readers too.

mandykqh committed:
Merge branch 'Fix-UI'

mandykqh committed:
Merge branch 'master' into branch-A-Assertions

mandykqh committed:
Merge pull request #3 from mandykqh/branch-A-Assertions

Implement A-Assertions

mandykqh committed:
Merge branch 'master' of https://github.com/mandykqh/ip

mandykqh committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/genie/command/AddCommand.java
#	src/main/java/genie/main/Parser.java
#	src/main/java/genie/main/TaskList.java

mandykqh committed:
Fix compilation bugs from resolving merge conflicts

mandykqh committed:
Merge pull request #4 from mandykqh/branch-A-CodeQuality

Implement A-CodeQuality

mandykqh committed:
Implement C-Help

New users using the app are not familiar with the commands to use.

Adding a "help" command allows users to familiarize themselves with the
commands. The help message is also shown when Genie first launches.

First-time users are immediately given guidance, and new users are given
guidance at any point of time using the "help" command.

mandykqh committed:
Merge branch 'branch-C-Help'

mandykqh committed:
Build Fat JAR file

mandykqh committed:
Improve GUI

mandykqh committed:
Update font

mandykqh committed:
Improve dialog and message bar GUI

mandykqh committed:
Fix find feature bug

mandykqh committed:
Improve dialog box GUI

mandykqh committed:
Implement A-BetterGUI

mandykqh committed:
Merge branch 'GUI'

mandykqh committed:
Clean up resources and code

mandykqh committed:
Implement A-UserGuide

mandykqh committed:
Fix load .txt file bug and improve UX

Contents of .txt file disappears after reading.
Users with existing saved data cannot view their list when launching
the app.

Showing the list upon launching the app serves as a refresher for
existing users. When there're no contents in the .txt file, assume that
the user is new and shows the help message instead.

mandykqh committed:
Update Genie's responses

mandykqh committed:
Clean up code

mandykqh committed:
Add JavaDoc
